pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java

```diff
@@ -0,0 +1,779 @@
+            pubAsync(serviceUnit, next)
+                    .whenComplete((__, e) -> log(e, serviceUnit, data, next));
+        }
+    }
+
+    private void handleReleaseEvent(String serviceUnit, ServiceUnitStateData data) {
+
+        if (isTargetBroker(data.sourceBroker())) {
+            ServiceUnitStateData next = new ServiceUnitStateData(Owned, data.broker(), data.sourceBroker());
+            // TODO: when close, pass message to clients to connect to the new broker
+            closeServiceUnit(serviceUnit)
+                    .thenCompose(__ -> pubAsync(serviceUnit, next))
+                    .whenComplete((__, e) -> log(e, serviceUnit, data, next));
+        }
+    }
+
+    private void handleSplitEvent(String serviceUnit, ServiceUnitStateData data) {
+        if (isTargetBroker(data.broker())) {
+            splitServiceUnit(serviceUnit)
+                    .thenCompose(__ -> tombstoneAsync(serviceUnit))
+                    .whenComplete((__, e) -> log(e, serviceUnit, data, null));
+        }
+    }
+
+    private void handleTombstoneEvent(String serviceUnit) {
+        closeServiceUnit(serviceUnit)
+                .thenAccept(__ -> {
+                    var request = getOwnerRequests.remove(serviceUnit);
+                    if (request != null) {
+                        request.completeExceptionally(new IllegalStateException("The ownership has been unloaded. "
+                                + "No owner is found for serviceUnit: " + serviceUnit));
+                    }
+                })
+                .whenComplete((__, e) -> log(e, serviceUnit, null, null));
+    }
+
+    private CompletableFuture<MessageId> pubAsync(String serviceUnit, ServiceUnitStateData data) {
+        CompletableFuture<MessageId> future = new CompletableFuture<>();
+        try {
+            outstandingPubMessages.acquire();
+        } catch (InterruptedException e) {
+            log.warn("Interrupted while acquiring semaphore to publish a message for serviceUnit:{}, data:{}",
+                    serviceUnit, data);
+            future.completeExceptionally(e);
+            return future;
+        }
+        producer.newMessage()
+                .key(serviceUnit)
+                .value(data)
+                .sendAsync()
+                .whenComplete((messageId, e) -> {
+                    outstandingPubMessages.release();
+                    if (e != null) {
+                        future.completeExceptionally(e);
+                    } else {
+                        future.complete(messageId);
+                    }
+                });
+        return future;
+    }
+
+    private CompletableFuture<MessageId> tombstoneAsync(String serviceUnit) {
+        return pubAsync(serviceUnit, null);
+    }
+
+    private boolean isTargetBroker(String broker) {
+        if (broker == null) {
+            return false;
+        }
+        // TODO: remove broker port from the input broker
```

Comment:
we should not remove the port, otherwise you won't be able to execute two brokers on the same host (I do it frequently  with Pulsar standalone)

Reply:
Removed the comment.